feat: add JSON type support

jchavarri opened the pull request:
- Add MYSQL_TYPE_JSON constant to FFI bindings
- Add Json variant to buffer_type and value types
- Implement json binding and field accessor functions
- Add comprehensive JSON tests for various data types
- Support both native JSON and string compatibility modes
- Tested against TiDB and MySQL 8.0.23

jchavarri committed:
feat: add JSON type support

- Add MYSQL_TYPE_JSON constant to FFI bindings
- Add Json variant to buffer_type and value types
- Implement json binding and field accessor functions
- Add comprehensive JSON tests for various data types
- Support both native JSON and string compatibility modes
- Tested against TiDB and MySQL 8.0.23

paurkedal committed:
Remove end-of-line spaces.

paurkedal commented on lib/field.ml:
```diff
@@ -9,6 +9,7 @@ type value =
   | `String of string
   | `Bytes of bytes
   | `Time of Time.t
+  | `Json of string
```
Would it not be better to use the `` `String`` case for JSON, as well?  I didn't write the original code, but the intention seem to me that the `value` type only contains the cases which have different representation on the OCaml side, e.g. omitting the various integer variants.

paurkedal commented on lib/field.ml:
```diff
@@ -124,6 +126,11 @@ let time field =
   | `Time t -> t
   | _ -> err field ~info:"a time value"
 
+let json field =
+  match value field with
+  | `Json j -> j
+  | _ -> err field ~info:"a json value"
+
```
In the test case you take into account the possibility that the database server returns a string (or bytes, as I've when testing).  That would be needed here, as well.  But if you agree to the above, this definition can be omitted.

paurkedal commented on lib/field.ml:
```diff
@@ -153,3 +160,9 @@ let time_opt field =
   | `Null -> None
   | _ -> err field ~info:"a nullable time value"
+
+let json_opt field =
+  match value field with
+  | `Json j -> Some j
+  | `Null -> None
+  | _ -> err field ~info:"a nullable json value"
```
Ditto.

paurkedal commented on tests/nonblocking/nonblocking_testsuite.ml:
```diff
@@ -345,10 +348,102 @@ struct
+          assert (Array.length row = 2);
+          (* Test that we can access the JSON field using different methods *)
+          let json_value = match M.Field.value row.(1) with
+            | `Json j -> j
+            | `String s -> s  (* TiDB/MySQL might return as string *)
```
I'm also seeing `` `Bytes`` here, using MariaDB client/server version 10.11.13.

paurkedal reviewed (changes requested):
Thanks, it's good to cover the JSON SQL type, though I don't think it needs to be exposed in the interface, see my comment below.

jongleb committed:
get rid of `Json constr

jchavarri committed:
Merge pull request #1 from jongleb/support-json

get rid of `Json constr

jchavarri committed:
refine test

jchavarri committed:
cleanup

jchavarri commented:
@paurkedal Thanks for the review, I think all the comments were tackled. Please let me know if anything else is missing.

paurkedal committed:
Also expect bytes when decoding JSON fields.

paurkedal commented:
I added a `` `Bytes`` case to `Field.sting` and `Field.string_opt`, since as we've seen returned when selecting the `data` JSON column, but on second though I'm not completely happy about this solution.  First of all, it breaks the currently strict interpretation of `Field.string` by allowing it to also convert from blob types, even if we only intended to add JSON.  Wondering why we're receiving bytes instead of a string, I found the answe in [MariaDB's explanation of the JSON type][1], which they implement as an alias for `LONGTEXT COLLATE utf8mb4_bin`.  Further, [`LONGTEXT` is classified together with blobs][2], so I think `` `Bytes`` is the correct type to expect for MariaDB.  I'm therefore considering whether we should return JSON as bytes.  And for consistency with the existing code, maybe restore `` `Json` as a `Bind.buffer_type` and adding it `to_blob`.

I need to think about it a bit more.  Let me know if you have any thoughts.

[1]: https://mariadb.com/docs/server/reference/data-types/string-data-types/json
[2]: https://mariadb.com/docs/server/reference/data-types/string-data-types/blob-and-text-data-types

paurkedal committed:
Use bytes for JSON objects for consistency with MariaDB.

This also reintroduce the `JSON case internally, making the code
consistent with other cases.

paurkedal commented:
Can you take a look at my latest commit to see if you agree with my implementation?  This follows the current pattern for other SQL types, but it means we receive the JSON as `bytes` instead of `string`.  I think the API documentation could use a table showing the correspondence between SQL and OCaml types, but I can do that after merging.

Shall we preserving the current commit history when I merge?
